File: e2e/examples_test.go
//go:build e2e

package e2e

import (
	"bytes"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
)

type exampleSpec struct {
	name          string
	secretConfigs map[string]string
	skipDestroy   bool
}

var exampleSpecs = []exampleSpec{
	{name: "multi-resource"},
	{name: "secrets-heavy", secretConfigs: map[string]string{"dbPassword": "test-pass-123"}},
	{name: "component"},
	{name: "replace-triggers"},
	{name: "large-state"},
	{name: "protect", skipDestroy: true},
	// stack-ref is excluded: it requires multi-resource deployed on a specific stack name simultaneously.
}

func TestExamples(t *testing.T) {
	truncateDB(t)

	sharedHome := t.TempDir()
	examplePulumi(t, t.TempDir(), sharedHome, "login", strataURL)

	for _, spec := range exampleSpecs {
		t.Run(spec.name, func(t *testing.T) {
			runExampleLifecycle(t, sharedHome, spec)
		})
	}
}

func runExampleLifecycle(t *testing.T, pulumiHome string, spec exampleSpec) {
	t.Helper()

	srcDir := filepath.Join(projectRoot, "examples", spec.name)
	workDir := copyExampleDir(t, srcDir)
	goModTidy(t, workDir)

	run := func(args ...string) (string, string) {
		return examplePulumi(t, workDir, pulumiHome, args...)
	}

	stackFQN := devOrgLogin + "/" + spec.name + "/e2e"

	run("stack", "init", "--stack", stackFQN)

	for key, val := range spec.secretConfigs {
		run("config", "set", "--secret", "--stack", stackFQN, key, val)
	}

	run("up", "--yes", "--stack", stackFQN)
	run("preview", "--stack", stackFQN)
	run("refresh", "--yes", "--stack", stackFQN)

	stdout, _ := run("stack", "export", "--stack", stackFQN)
	exportFile := filepath.Join(t.TempDir(), "state.json")
	if err := os.WriteFile(exportFile, []byte(stdout), 0o644); err != nil {
		t.Fatalf("write export file: %v", err)
	}
	run("stack", "import", "--stack", stackFQN, "--file", exportFile)

	if !spec.skipDestroy {
		run("destroy", "--yes", "--stack", stackFQN)
	}

	run("stack", "rm", "--yes", "--force", "--stack", stackFQN)
}

func examplePulumi(t *testing.T, dir, pulumiHome string, args ...string) (string, string) {
	t.Helper()

	cmd := exec.Command("pulumi", args...)
	cmd.Dir = dir
	cmd.Env = pulumiEnv(pulumiHome)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		t.Fatalf("pulumi %s failed: %v\nstdout:\n%s\nstderr:\n%s",
			strings.Join(args, " "), err, stdout.String(), stderr.String())
	}
	return stdout.String(), stderr.String()
}

func pulumiEnv(pulumiHome string) []string {
	overrides := map[string]string{
		"PULUMI_HOME":              pulumiHome,
		"PULUMI_ACCESS_TOKEN":      devAuthToken,
		"PULUMI_SKIP_UPDATE_CHECK": "true",
		"PULUMI_CONFIG_PASSPHRASE": "test",
		"PULUMI_DIY_BACKEND_URL":   "",
		"PULUMI_BACKEND_URL":       "",
	}

	var env []string
	for _, e := range os.Environ() {
		key, _, _ := strings.Cut(e, "=")
		if _, ok := overrides[key]; !ok {
			env = append(env, e)
		}
	}
	for k, v := range overrides {
		env = append(env, k+"="+v)
	}
	return env
}

func goModTidy(t *testing.T, dir string) {
	t.Helper()
	cmd := exec.Command("go", "mod", "tidy")
	cmd.Dir = dir
	cmd.Env = os.Environ()
	if out, err := cmd.CombinedOutput(); err != nil {
		t.Fatalf("go mod tidy in %s: %v\n%s", filepath.Base(dir), err, out)
	}
}

func copyExampleDir(t *testing.T, src string) string {
	t.Helper()
	dst := t.TempDir()

	entries, err := os.ReadDir(src)
	if err != nil {
		t.Fatalf("read example dir %s: %v", src, err)
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		srcPath := filepath.Join(src, entry.Name())
		dstPath := filepath.Join(dst, entry.Name())
		cpFile(t, srcPath, dstPath)
	}

	return dst
}

func cpFile(t *testing.T, src, dst string) {
	t.Helper()
	in, err := os.Open(src)
	if err != nil {
		t.Fatalf("open %s: %v", src, err)
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		t.Fatalf("create %s: %v", dst, err)
	}
	defer out.Close()

	if _, err := io.Copy(out, in); err != nil {
		t.Fatalf("copy %s: %v", filepath.Base(src), err)
	}
}

File: e2e/examples.test.ts
// E2E — Go example programs: full lifecycle for each example.

import { afterAll, beforeAll, describe, expect, setDefaultTimeout, test } from "bun:test";
import {
	apiRequest,
	BACKEND_URL,
	cleanupDir,
	copyExampleDir,
	createPulumiHome,
	pulumi,
	truncateTables,
} from "./helpers.js";
import "./setup.js";

// Examples take time: Go compilation + provider plugin downloads.
setDefaultTimeout(180_000);

let pulumiHome: string;

beforeAll(async () => {
	pulumiHome = await createPulumiHome();
	await pulumi(["login", "--cloud-url", BACKEND_URL], { pulumiHome });
});

afterAll(async () => {
	await cleanupDir(pulumiHome);
	await truncateTables();
});

// ============================================================================
// Helper: run the standard full lifecycle for a Go example
// ============================================================================

function runExampleLifecycle(
	exampleName: string,
	opts: {
		configSteps?: Array<{ key: string; value: string }>;
		skipDestroyRm?: boolean;
	} = {},
): void {
	const stackFQN = `dev-org/${exampleName}/dev`;
	const stackPath = `/stacks/dev-org/${exampleName}/dev`;
	let dir: string;

	beforeAll(async () => {
		dir = await copyExampleDir(exampleName);
	});

	afterAll(async () => {
		if (dir) await cleanupDir(dir);
	});

	test("stack init", async () => {
		const res = await pulumi(["stack", "init", stackFQN], { cwd: dir, pulumiHome });
		expect(res.exitCode).toBe(0);
	});

	for (const { key, value } of opts.configSteps ?? []) {
		test(`config set ${key}`, async () => {
			const res = await pulumi(["config", "set", key, value], { cwd: dir, pulumiHome });
			expect(res.exitCode).toBe(0);
		});
	}

	test("pulumi up", async () => {
		const res = await pulumi(["up", "--yes"], { cwd: dir, pulumiHome });
		expect(res.exitCode).toBe(0);
	});

	test("pulumi preview", async () => {
		const res = await pulumi(["preview"], { cwd: dir, pulumiHome });
		expect(res.exitCode).toBe(0);
	});

	test("pulumi refresh", async () => {
		const res = await pulumi(["refresh", "--yes"], { cwd: dir, pulumiHome });
		expect(res.exitCode).toBe(0);
	});

	test("export returns resources", async () => {
		const res = await apiRequest(`${stackPath}/export`);
		expect(res.status).toBe(200);
		const body = await res.json();
		expect(body.deployment).toBeDefined();
	});

	test("import roundtrip", async () => {
		const exportRes = await apiRequest(`${stackPath}/export`);
		const deployment = await exportRes.json();
		const importRes = await apiRequest(`${stackPath}/import`, {
			method: "POST",
			body: deployment,
		});
		expect(importRes.status).toBe(200);
		const result = await importRes.json();
		expect(result).toHaveProperty("updateId");
	});

	if (!opts.skipDestroyRm) {
		test("pulumi destroy", async () => {
			const res = await pulumi(["destroy", "--yes"], { cwd: dir, pulumiHome });
			expect(res.exitCode).toBe(0);
		});

		test("pulumi stack rm", async () => {
			const res = await pulumi(["stack", "rm", "--yes"], { cwd: dir, pulumiHome });
			expect(res.exitCode).toBe(0);
		});
	}
}

// ============================================================================
// Examples
// ============================================================================

describe("example: multi-resource", () => {
	runExampleLifecycle("multi-resource");
});

describe("example: secrets-heavy", () => {
	runExampleLifecycle("secrets-heavy", {
		configSteps: [{ key: "dbHost", value: "testdb.example.com" }],
	});
});

describe("example: component", () => {
	runExampleLifecycle("component");
});

describe("example: replace-triggers", () => {
	runExampleLifecycle("replace-triggers");
});

describe("example: large-state", () => {
	runExampleLifecycle("large-state");
});

describe("example: protect", () => {
	// Protect flag blocks destroy by design — skip destroy and rm.
	// Stack is cleaned up via truncateTables() in afterAll.
	runExampleLifecycle("protect", { skipDestroyRm: true });
});
